packages/go_router/lib/src/go_router.dart

```diff
@@ -113,8 +137,18 @@ class GoRouter extends ChangeNotifier with NavigatorObserver {
 
   /// Push a URI location onto the page stack w/ optional query parameters, e.g.
   /// `/family/f2/person/p1?color=blue`
-  void push(String location, {Object? extra}) =>
-      routerDelegate.push(location, extra: extra);
+  void push(String location, {Object? extra}) {
+    assert(() {
+      log.info('pushing $location');
+      return true;
+    }());
+    routeInformationParser
+        .parseRouteInformation(
+            DebugGoRouteInformation(location: location, state: extra))
+        .then<void>((List<GoRouteMatch> matches) {
```

Comment:
I'm a bit concerned by this `.then`. Today this keeps the existing behavior as `parseRouteInformation` returns a `SynchronousFuture`. However if for whatever reason `parseRouteInformation` ever return an asynchronous future, now `push` would return before the location has been pushed.

Is this a valid concern? How can we mitigate it? Should this method assert that `parseRouteInformation` returned a `SynchornousFuture`?

Reply:
This is a good point, If we choose to make redirects asynchronous this will be a concern. For now, I created https://github.com/flutter/flutter/issues/105808 to track because I think this is something we should support.

You're right we shouldn't hide the fact that push() can be asynchronous, so we might want to change the return type for push() to `Future<void>`.

We could leave it as-is and follow up when we get to https://github.com/flutter/flutter/issues/105808. This PR already contains breaking changes so it's not a bad time to break things either.